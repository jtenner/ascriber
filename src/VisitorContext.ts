import { Node } from "assemblyscript";

export class VisitorContext<T extends Node> {
  public constructor(
    public node: T,
    public parentContext: VisitorContext<Node> | null = null,
  ) {}
}
