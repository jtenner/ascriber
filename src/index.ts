export * from "./api";
export * from "./Visitor";
export * from "./VisitorContext";
export * as cli from "./cli";
