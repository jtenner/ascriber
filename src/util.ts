import { Parser, Program, Source, Expression } from "assemblyscript";
import { IVisitorObject, Visitor } from "./Visitor";

export interface ITemplateObject {
  [name: string]: Expression;
}

export function visit(source: string, query: Partial<IVisitorObject>): Source {
  const visitor = new Visitor();
  const result = parse(source);
  visitor.traverse(query, result);
  return result;
}

export function parse(source: string): Source {
  const parser = new Parser(({
    diagnostics: [],
    sources: [],
  } as unknown) as Program);
  parser.parseFile(source, "source.ts", true);
  return parser.currentSource;
}

/*
export function template(source: string, templateObject: ITemplateObject): Source {

}
*/
