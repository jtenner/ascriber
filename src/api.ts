import { Parser, Source, Expression } from "assemblyscript";
import { IVisitorObject, Visitor } from "./Visitor";
import { ASTBuilder } from "./util/ast";

export interface ITemplateObject {
  [name: string]: Expression;
}

export function visit(source: string, query: Partial<IVisitorObject>): Source {
  const visitor = new Visitor();
  const result = parse(source);
  visitor.traverse(query, result);
  return result;
}

export function parse(source: string): Source {
  const diagnostics = {
    diagnostics: [],
    sources: [],
  } as any;
  const parser = new Parser(diagnostics);
  parser.parseFile(source, "source.ts", true);
  return parser.currentSource;
}

export function transform(
  source: string,
  query: Partial<IVisitorObject>,
): string {
  const output = visit(source, query);
  const builder = new ASTBuilder();
  builder.visitSource(output);
  return builder.finish();
}

/*
export function template(source: string, templateObject: ITemplateObject): Source {

}
*/
