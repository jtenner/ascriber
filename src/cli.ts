import { parse, Config, help } from "./util/parser";
import colors from "ansi-colors";
const pkg = require("../package.json");

const config: Config = {
  help: {
    alias: "h",
    default: false,
    type: "b",
    description: "Show this help screen.",
  },
};

export enum ExitCode {
  Success = 0,
  Fail = 0,
}

export interface IWritable {
  write(value: string): void;
}

export interface IExitable {
  exit(code: ExitCode): void;
}

export interface ICLIConfiguration {
  stdout: IWritable;
  stderr: IWritable;
  process: IExitable;
}

export function main(argv: string[], cliConfig: ICLIConfiguration): void {
  const cliResult = parse(argv, config);
  const options = cliResult.options;
  const unknown = cliResult.unknown;
  // const entryPoints = cliResult.args;
  // const rest = cliResult.trailing;

  const { stderr, stdout, process } = cliConfig;

  stdout.write(`ascriber cli version ${pkg.version}\n`);

  if (unknown.length > 0) {
    for (const arg of unknown) {
      stderr.write(`Unknown cli option: ${colors.red(arg)}\n`);
    }
    return process.exit(1);
  }

  if (options.help) {
    const helpText = help(config);
    stdout.write(`\n${helpText}\n`);
    return process.exit(0);
  }

  return process.exit(1);
}
