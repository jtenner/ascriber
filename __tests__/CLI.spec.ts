import { cli } from "../src";

const argList: string[][] = [
  ["long help", "--help"],
  ["short help", "-h"],
  ["invalid", "--invalid"],
];

interface IMockExitable extends cli.IExitable {
  code: number;
}

interface IMockWritable extends cli.IWritable {
  output: string;
}

function createMockExitable(): IMockExitable {
  const result = {
    code: 0,
    exit(code: number): void {
      result.code = code;
    },
  };
  return result;
}

function createMockWritable(): IMockWritable {
  const result = {
    output: "",
    write(contents: string): void {
      result.output += contents;
    },
  };
  return result;
}

function createMockCLIConfig() {
  return {
    process: createMockExitable(),
    stdout: createMockWritable(),
    stderr: createMockWritable(),
  };
}

test("cliOutput", () => {
  for (const [name, ...rest] of argList) {
    const config = createMockCLIConfig();
    cli.main(rest, config);
    expect(config.process.code).toMatchSnapshot(`${name} code`);
    expect(config.stdout.output).toMatchSnapshot(`${name} stdout`);
    expect(config.stderr.output).toMatchSnapshot(`${name} stderr`);
  }
});
