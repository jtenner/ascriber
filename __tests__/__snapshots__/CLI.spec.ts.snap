// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cliOutput: invalid code 1`] = `1`;

exports[`cliOutput: invalid stderr 1`] = `
"Unknown cli option: [31m--invalid[39m
"
`;

exports[`cliOutput: invalid stdout 1`] = `
"ascriber cli version 0.0.1
"
`;

exports[`cliOutput: long help code 1`] = `0`;

exports[`cliOutput: long help stderr 1`] = `""`;

exports[`cliOutput: long help stdout 1`] = `
"ascriber cli version 0.0.1

  --help, -h            Show this help screen.
"
`;

exports[`cliOutput: short help code 1`] = `0`;

exports[`cliOutput: short help stderr 1`] = `""`;

exports[`cliOutput: short help stdout 1`] = `
"ascriber cli version 0.0.1

  --help, -h            Show this help screen.
"
`;
